package/MDAnalysis/lib/c_distances.pyx

```diff
@@ -315,3 +317,290 @@ def contact_matrix_pbc(coord, sparse_contacts, box, cutoff):
+                
+                
+                if lo_z:
+                    for j in range(3):
+                        output[p, j] = coord[j] - shiftX[j] - shiftY[j] + shiftZ[j]
+                    indices[p] = i
+                    p += 1
+                    
+                elif hi_z:
+                    for j in range(3):
+                        output[p, j] = coord[j] - shiftX[j] - shiftY[j] - shiftZ[j]
+                    indices[p] = i
+                    p += 1
+                    
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftX[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftX[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+        if lo_y:
+            for j in range(3):
+                output[p, j] = coord[j] + shiftY[j]
+            indices[p] = i
+            p += 1
+            
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] + shiftY[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] + shiftY[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+        elif hi_y:
+            for j in range(3):
+                output[p, j] = coord[j] - shiftY[j]
+            indices[p] = i
+            p += 1
+            
+
+            if lo_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftY[j] + shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+            elif hi_z:
+                for j in range(3):
+                    output[p, j] = coord[j] - shiftY[j] - shiftZ[j]
+                indices[p] = i
+                p += 1
+                
+        if lo_z:
+            for j in range(3):
+                output[p, j] = coord[j] + shiftZ[j]
+            indices[p] = i
+            p += 1
+            
+        elif hi_z:
+            for j in range(3):
+                output[p, j] = coord[j] - shiftZ[j]
+            indices[p] = i
+            p += 1
+            
+    return numpy.array(output[:p]), numpy.array(indices[:p])
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef float _dot(float  *a, float *b):
+    """Return dot product of two sequences in range."""
+    cdef ssize_t n
+    cdef float sum1
+    cdef ssize_t dim
+    
+    dim=3
+    sum1 = 0.0
+    for n in range(dim):
+        sum1 += a[n] * b[n]
+    return sum1
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def undo_augment(int[:] results, int[:] translation, int nreal):
+    """Translate augmented indices back to originals
+    
+    Note: modifies results in place!
```

Comment:
Actually, I thought for complex selections it might be useful to keep all the particles (original + images). But it looks like its better to modify the results, otherwise it will lead to multiple computations of similar points with increase in number of selections(operations). 
Meanwhile, I will also try to think of a  case, where it might fail.